libraries/chain/webassembly/eos-vm-oc/code_cache.cpp

```diff
@@ -0,0 +1,376 @@
+   if(dirty_region.flush(0, 0, false) == false)
+      elog("Syncing code cache failed");
+}
+
+template <typename T>
+void code_cache_base::serialize_cache_index(fc::datastream<T>& ds) {
+   unsigned entries = _cache_index.size();
+   fc::raw::pack(ds, entries);
+   for(const code_descriptor& cd : _cache_index)
+      fc::raw::pack(ds, cd);
+}
+
+code_cache_base::~code_cache_base() {
+   //reopen the code cache in our process
+   struct stat st;
+   if(fstat(_cache_fd, &st))
+      return;
+   char* code_mapping = (char*)mmap(nullptr, st.st_size, PROT_READ|PROT_WRITE, MAP_SHARED, _cache_fd, 0);
+   if(code_mapping == MAP_FAILED)
+      return;
+
+   allocator_t* allocator = reinterpret_cast<allocator_t*>(code_mapping);
+
+   //serialize out the cache index
+   fc::datastream<size_t> dssz;
+   serialize_cache_index(dssz);
+   size_t sz = dssz.tellp();
+
+   char* p = nullptr;
+   while(_cache_index.size()) {
+      p = (char*)allocator->allocate(sz);
+      if(p != nullptr)
+         break;
+      //in theory, there could be too little free space avaiable to store the cache index
+      //try to free up some space
+      for(unsigned int i = 0; i < 25 && _cache_index.size(); ++i) {
+         allocator->deallocate(code_mapping + _cache_index.back().code_begin);
+         allocator->deallocate(code_mapping + _cache_index.back().initdata_begin);
+         _cache_index.pop_back();
+      }
+   }
+
+   if(p) {
+      fc::datastream<char*> ds(p, sz);
+      serialize_cache_index(ds);
+
+      uintptr_t ptr_offset = p-code_mapping;
+      *((uintptr_t*)(code_mapping+descriptor_ptr_from_file_start)) = ptr_offset;
+   }
+   else
+      *((uintptr_t*)(code_mapping+descriptor_ptr_from_file_start)) = 0;
+
+   msync(code_mapping, allocator->get_size(), MS_SYNC);
+   munmap(code_mapping, allocator->get_size());
+   close(_cache_fd);
+   set_on_disk_region_dirty(false);
+
+}
+
+void code_cache_base::free_code(const digest_type& code_id, const uint8_t& vm_version) {
+   code_cache_index::index<by_hash>::type::iterator it = _cache_index.get<by_hash>().find(boost::make_tuple(code_id, vm_version));
+   if(it != _cache_index.get<by_hash>().end()) {
+      write_message_with_fds(_compile_monitor_write_socket, evict_wasms_message{ {*it} });
+      _cache_index.get<by_hash>().erase(it);
+   }
+
+   //if it's in the queued list, erase it
+   _queued_compiles.erase({code_id, vm_version});
+
+   //however, if it's currently being compiled there is no way to cancel the compile,
+   //so instead set a poison boolean that indicates not to insert the code in to the cache
+   //once the compile is complete
+   const std::unordered_map<code_tuple, bool>::iterator compiling_it = _outstanding_compiles_and_poison.find({code_id, vm_version});
+   if(compiling_it != _outstanding_compiles_and_poison.end())
+      compiling_it->second = true;
+}
+
+void code_cache_base::check_eviction_threshold(size_t free_bytes) {
+   if(free_bytes < _free_bytes_eviction_threshold) {
```

Comment:
What if the cache is fragmented due to many small wasms?  Can we get in a situation where nothing is ever evicted because we have enough free bytes in total, but the wasms we want to execute don't fit in any free segment?

Reply:
Maybe on a toofull message from the compile process, cache eviction should be more aggressive and disregard the threshold